overlord/snapstate/handlers_components_download_test.go

```diff
@@ -0,0 +1,172 @@
+package snapstate_test
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/snapstate/snapstatetest"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/store/storetest"
+
+	. "gopkg.in/check.v1"
+)
+
+type downloadComponentSuite struct {
+	baseHandlerSuite
+
+	fakeStore *fakeStore
+}
+
+var _ = Suite(&downloadComponentSuite{})
+
+func (s *downloadComponentSuite) SetUpTest(c *C) {
+	s.baseHandlerSuite.SetUpTest(c)
+
+	s.fakeStore = &fakeStore{
+		state:       s.state,
+		fakeBackend: s.fakeBackend,
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	s.AddCleanup(snapstatetest.UseFallbackDeviceModel())
+
+	snapstate.ReplaceStore(s.state, s.fakeStore)
+	s.state.Set("refresh-privacy-key", "privacy-key")
+}
+
+func (s *downloadComponentSuite) TestDoDownloadComponentNormal(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	si := &snap.SideInfo{
+		RealName: "snap",
+		SnapID:   snaptest.AssertedSnapID("snap"),
+		Revision: snap.R(11),
+		Channel:  "latest/stable",
+	}
+
+	t := s.state.NewTask("download-component", "...")
+
+	t.Set("snap-setup", &snapstate.SnapSetup{
```

Comment:
The rate limiting isn't tested, there could be a simple variant of this with `AutoRefresh=true` that checks that

Reply:
Thanks, added in a4ee50ac0ccb1c6f21300936fd6dbf3402e904d5.